common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -1124,9 +1143,23 @@ private boolean isDuplicateBlock() {
      * block parts buffered in memory.
      */
     private void writeDeferredBufs() throws IOException {
+      long totalSize = 0;
       for (ByteBuffer deferredBuf : deferredBufs) {
+        totalSize += deferredBuf.limit();
         writeBuf(deferredBuf);
       }
+      mergeManager.pushMergeMetrics.cachedBlockBytes.dec(totalSize);
+      deferredBufs = null;
+    }
+
+    private void freeDeferredBufs() {
+      if (deferredBufs != null && !deferredBufs.isEmpty()) {
+        long totalSize = 0;
+        for (ByteBuffer deferredBuf : deferredBufs) {
+          totalSize += deferredBuf.limit();
+        }
+        mergeManager.pushMergeMetrics.cachedBlockBytes.dec(totalSize);
+      }
       deferredBufs = null;
```

Comment:
There are a bunch of other places where we simply drop `deferredBufs` entirely (`abortIfNecessary`, `onData`, etc).
Given this is tracked as a counter, we should update there as well (since we have already `inc` on it).

Essentially, replace all `deferredBufs = null` with `freeDeferredBufs`

Reply:
Makes sense. Have checked all the places and made the replacement, it seems just applies to `abortIfNecessary` and `onData`.

Reply:
Makes sense. Have checked all the places and made the replacement, looks like the remaining places that need it are only `abortIfNecessary` and `onData`.